### PR TITLE
Ch112/create addtrader command

### PR DESCRIPTION
**Pull request by Darkvodou**

### CODE REVIEW

File: src/main/java/fr/darkvodou/NKtrader/NKtrader.java [hunk 1]
```diff
@@ -1,18 +1,24 @@
 package fr.darkvodou.NKtrader;
 
 import fr.darkvodou.NKtrader.cmds.addTraderCmd;
+import fr.darkvodou.NKtrader.listeners.LeftClick;
 import fr.darkvodou.NKtrader.managers.Manager;
 import org.bukkit.Bukkit;
+import org.bukkit.Material;
 import org.bukkit.command.ConsoleCommandSender;
+import org.bukkit.event.Listener;
 import org.bukkit.plugin.java.JavaPlugin;
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/NKtrader.java]: remove space

File: src/main/java/fr/darkvodou/NKtrader/NKtrader.java [hunk 2]
```diff
@@ -30,10 +36,23 @@ public void onEnable()
 			return;
 		}
 
+		//create event
+
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/NKtrader.java]: remove space

File: src/main/java/fr/darkvodou/NKtrader/NKtrader.java [hunk 3]
```diff
@@ -1,18 +1,24 @@
...
+import java.util.ArrayList;
+
 import static fr.darkvodou.NKtrader.enums.MsgUtils.PREFIX_SUCCESS;
 
 public final class NKtrader extends JavaPlugin
 {
 	private static NKtrader plugin = null;
 	private ConsoleCommandSender console = null;
 	private Manager manager = null;
+	private ArrayList<Listener> listeners = new ArrayList<>();
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/NKtrader.java]: private **List**<Listener> listeners = new ArrayList<>();

File: src/main/java/fr/darkvodou/NKtrader/NKtrader.java [hunk 4]
```diff
@@ -30,10 +36,23 @@ public void onEnable()
 			return;
 		}
 
+		//create event
+
+		Listener LeftClick = new LeftClick(manager.getTraderManager());
+		listeners.add(LeftClick);
+
+		//register event
+
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/NKtrader.java]: remove space

File: src/main/java/fr/darkvodou/NKtrader/NKtrader.java [hunk 5]
```diff
@@ -30,10 +36,23 @@ public void onEnable()
...
+		for(Listener event : listeners)
+		{
+			getServer().getPluginManager().registerEvents(event,this);
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/NKtrader.java]: CTRL ALT L

File: src/main/java/fr/darkvodou/NKtrader/NKtrader.java [hunk 6]
```diff
@@ -30,10 +36,23 @@ public void onEnable()
 			return;
 		}
 
+		//create event
+
+		Listener LeftClick = new LeftClick(manager.getTraderManager());
+		listeners.add(LeftClick);
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/NKtrader.java]: listeners.add(new LeftClick(manager.getTraderManager()));

File: src/main/java/fr/darkvodou/NKtrader/NKtrader.java [hunk 7]
```diff
@@ -30,10 +36,23 @@ public void onEnable()
...
+			getServer().getPluginManager().registerEvents(event,this);
+		}
+
 		//Register commands
-		getCommand("addtrader").setExecutor(new addTraderCmd());
+		getCommand("addtrader").setExecutor(new addTraderCmd(manager));
 
-		console.sendMessage(PREFIX_SUCCESS + " :Ready !");
+		console.sendMessage(PREFIX_SUCCESS + ": Ready !");
+		console.sendMessage(PREFIX_SUCCESS + ": " + Material.CHEST.toString());
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/NKtrader.java]: useless

File: src/main/java/fr/darkvodou/NKtrader/enums/MsgUtils.java [hunk 8]
```diff
@@ -18,7 +21,13 @@
 	ERROR_CONSOL_IS_SENDER(PREFIX_ERROR + "Vous ne pouvez pas executer cette commande (console ou autre)."),
 	ERROR_HAS_NOT_PERM(PREFIX_ERROR + "Vous n'avez pas la permission d'executer cette commande."),
 	ERROR_NOT_ENOUGH_ARGS(PREFIX_ERROR + "Vous n'avez pas mis assez d'arguments."),
-	ERROR_BAD_ARGS(PREFIX_ERROR + "Un ou plusieurs argument(s) ne sont pas correcte(s), verifiez la syntaxe.");
+	ERROR_BAD_ARGS(PREFIX_ERROR + "Un ou plusieurs argument(s) ne sont pas correcte(s), verifiez la syntaxe."),
+	ERROR_TOO_MANY_ARGS(PREFIX_ERROR + "Erreur syntax, trop d'arguments"),
+	ERROR_TRY_KILL_TRADER(PREFIX_ERROR + "Hey ! Pas gentil ça !"),
+	ERROR_NAME_LENGTH(PREFIX_ERROR + "Nom trop long (100 max)"),
+	ERROR_ENTITY_TYPE(PREFIX_ERROR + "L'entityType précisé n'existe pas"),
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/enums/MsgUtils.java]: user friendly

File: src/main/java/fr/darkvodou/NKtrader/enums/MsgUtils.java [hunk 9]
```diff
@@ -18,7 +21,13 @@
...
+	ERROR_MATERIAL_TYPE(PREFIX_ERROR + "Le materialType précisé n'existe pas"),
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/enums/MsgUtils.java]: user friendly

File: src/main/java/fr/darkvodou/NKtrader/enums/MsgUtils.java [hunk 10]
```diff
@@ -18,7 +21,13 @@
...
+	ERROR_WORLD_NOT_EXIST(PREFIX_ERROR + "Le monde que vous voulez spécifier semble ne pas exister");
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/enums/MsgUtils.java]: voulez spécifier => avez spécifié

File: src/main/java/fr/darkvodou/NKtrader/managers/TraderManager.java [hunk 11]
```diff
@@ -36,15 +36,15 @@ public boolean hasTrader(Location location)
 		return traders.containsKey(location.getWorld().getName() + location.getBlockX() + location.getBlockY() + location.getBlockZ());
 	}
 
-	public void addTrader(Trader trader)
+	public boolean addTrader(Trader trader)
 	{
 		String id = trader.getId();
 
 		if(hasTrader(id))
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/managers/TraderManager.java]: hasTrader => existTrader

File: src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java [hunk 12]
```diff
@@ -1,19 +1,39 @@
...
+import static fr.darkvodou.NKtrader.enums.Msgmanager.ERROR_TRADER_IS_CONTAINED;
+import static fr.darkvodou.NKtrader.enums.Msgmanager.ERROR_TRADER_IS_SET;
 import static fr.darkvodou.NKtrader.enums.Permissions.*;
 import static fr.darkvodou.NKtrader.enums.Usages.*;
+import static fr.darkvodou.NKtrader.utils.CheckType.isNumber;
+import static java.lang.Double.parseDouble;
 
 public class addTraderCmd implements CommandExecutor
 {
+	private final Manager manager;
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java]: Use TraderManager instead

File: src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java [hunk 13]
```diff
@@ -1,19 +1,39 @@
...
+		this.manager = manager;
+	}
+
 	@Override
 	public boolean onCommand(CommandSender sender, org.bukkit.command.Command command, String label, String[] args)
 	{
+		String type = "";
+		String name = "";
+		Location location;
+		World world = Bukkit.getWorld("world");
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java]: World world;

File: src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java [hunk 14]
```diff
@@ -1,19 +1,39 @@
...
+
+	public addTraderCmd(Manager manager)
+	{
+		this.manager = manager;
+	}
+
 	@Override
 	public boolean onCommand(CommandSender sender, org.bukkit.command.Command command, String label, String[] args)
 	{
+		String type = "";
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java]: move all var in "instance of Player" if

File: src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java [hunk 15]
```diff
@@ -30,27 +50,311 @@ public boolean onCommand(CommandSender sender, org.bukkit.command.Command comman
 
 			if(args.length == 0)
 			{
-				sender.sendMessage("" + NKT_USAGE_ADDTRADER);
+				sender.sendMessage("1 :" + NKT_USAGE_ADDTRADER);
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java]: debug?

File: src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java [hunk 16]
```diff
@@ -30,27 +50,311 @@ public boolean onCommand(CommandSender sender, org.bukkit.command.Command comman
...
+				int j = 0;
+				int i;
+
+				//If name composed increase nbArgs by the length of the composed name
+				if(args[nPlace + 1].charAt(0) == '"')
+				{
+
+					for(i = nPlace + 1; i < args.length && args[i].charAt(args[i].length() - 1) != '"'; i++)
+					{
+						j++;
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java]: build the name here

File: src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java [hunk 17]
```diff
@@ -30,27 +50,311 @@ public boolean onCommand(CommandSender sender, org.bukkit.command.Command comman
...
+					if(args[i].charAt(args[i].length() - 1) != '"')
+					{
+						sender.sendMessage(ERROR_BAD_ARGS + " 8:\n" + NKT_USAGE_ADDTRADER);
+
+						return true;
+					}
+
+					nbArgsForN = j;
+				}
+
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java]: else

File: src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java [hunk 18]
```diff
@@ -30,27 +50,311 @@ public boolean onCommand(CommandSender sender, org.bukkit.command.Command comman
...
+				if(args.length > tPlace + 2)
+				{
+					if(!args[tPlace + 2].equals("-n") && !args[tPlace + 2].equals("-w"))
+					{
+						sender.sendMessage(ERROR_BAD_ARGS + " 11:\n" + NKT_USAGE_ADDTRADER);
+
+						return true;
+					}
+				}
+
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java]: verify type here

File: src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java [hunk 19]
```diff
@@ -30,27 +50,311 @@ public boolean onCommand(CommandSender sender, org.bukkit.command.Command comman
...
+				location = new Location(world, parseDouble(args[wPlace + 2]), parseDouble(args[wPlace + 3]), parseDouble(args[wPlace + 4]));
+			}
+			else
+			{
+				location = ((Player) sender).getLocation();
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java]: get world

File: src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java [hunk 20]
```diff
@@ -30,27 +50,311 @@ public boolean onCommand(CommandSender sender, org.bukkit.command.Command comman
...
+				}
+
+				type = args[tPlace + 1];
+			}
+
+			Trader trader;
+
+			if(args[0].equals("entity"))
+			{
+				if(world == null)
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java]: useless

File: src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java [hunk 21]
```diff
@@ -30,27 +50,311 @@ public boolean onCommand(CommandSender sender, org.bukkit.command.Command comman
...
+				{
+					sender.sendMessage(ERROR_WORLD_NOT_EXIST + "");
+					return true;
+				}
+
+				EntityType entityType = EntityType.valueOf("villager".toUpperCase());
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java]: EntityType.VILLAGER

File: src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java [hunk 22]
```diff
@@ -30,27 +50,311 @@ public boolean onCommand(CommandSender sender, org.bukkit.command.Command comman
...
+
+				if(tPlace != -1)
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java]: if(!type.equals(""))

File: src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java (hunk 20 above)
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java]: Create methode "createTraderEntity(trader)" in TraderManager

File: src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java (hunk 14 above)
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java]: rename to dataType

File: src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java [hunk 23]
```diff
@@ -30,27 +50,311 @@ public boolean onCommand(CommandSender sender, org.bukkit.command.Command comman
...
+					entity.setCustomName(name);
+					entity.setCustomNameVisible(true);
+				}
+			}
+			else
+			{
+				Block block = location.getBlock();
+				Material material = Material.CHEST;
+
+				if(tPlace != -1)
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java]: if(!dataType.equals(""))

File: src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java [hunk 24]
```diff
@@ -30,27 +50,311 @@ public boolean onCommand(CommandSender sender, org.bukkit.command.Command comman
...
+					if(material == null)
+					{
+						sender.sendMessage(ERROR_MATERIAL_TYPE + "");
+
+						return true;
+					}
+					sender.sendMessage(DEBUG + block.getBlockData().toString());
+				}
+
+				if(block.getType().isAir())
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java]: useless check

File: src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java [hunk 25]
```diff
@@ -30,27 +50,311 @@ public boolean onCommand(CommandSender sender, org.bukkit.command.Command comman
...
 				}
 				else
 				{
-					//Block
+					trader = new Trader(location, args[0], location.getBlock().getType().toString());
+				}
+
+				if(!name.equals(""))
+				{
+					if(world == null)
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java]: useless

File: src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java [hunk 26]
```diff
@@ -30,27 +50,311 @@ public boolean onCommand(CommandSender sender, org.bukkit.command.Command comman
...
+					{
+						sender.sendMessage(ERROR_WORLD_NOT_EXIST + "");
+
+						return true;
+					}
+
+					Location loct = location.getBlock().getLocation();
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java]: more explicit var name

File: src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java [hunk 27]
```diff
@@ -30,27 +50,311 @@ public boolean onCommand(CommandSender sender, org.bukkit.command.Command comman
...
+					loct.add(0.5, 1.0, 0.5);
+					Entity armor_stand = world.spawnEntity(loct, EntityType.ARMOR_STAND);
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java]: camelCase

File: src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java [hunk 28]
```diff
@@ -30,27 +50,311 @@ public boolean onCommand(CommandSender sender, org.bukkit.command.Command comman
...
+
+					if(armor_stand instanceof LivingEntity)
+					{
+						LivingEntity livingEntity = (LivingEntity) armor_stand;
+						livingEntity.setInvisible(true);
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java]: add setInvincible

File: src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java [hunk 29]
```diff
@@ -30,27 +50,311 @@ public boolean onCommand(CommandSender sender, org.bukkit.command.Command comman
...
+
+					trader.setName(name);
+					armor_stand.setCustomName(name);
+					armor_stand.setCustomNameVisible(true);
 				}
 			}
+
+			if(!manager.getTraderManager().addTrader(trader))
+			{
+				sender.sendMessage(ERROR_TRADER_IS_CONTAINED + ", look at console");
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java]: add return true

File: src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java [hunk 30]
```diff
@@ -30,27 +50,311 @@ public boolean onCommand(CommandSender sender, org.bukkit.command.Command comman
...
+
+			i++;
+		}
+
+		return -1;
+	}
+
 	private boolean hasAdminPermissions(CommandSender sender)
 	{
 		return sender.hasPermission("" + NKT_ADMIN) || sender.hasPermission("" + NKT_ALL);
```
Comment (nokisenpai) [inline src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java]: check permission "NKTrader.*"

**Review by nokisenpai — CHANGES_REQUESTED** [review]

File: src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java (hunk 22 above)
Comment (Darkvodou) [reply src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java]: moved in a tPlace != -1 where dataType is defined

File: src/main/java/fr/darkvodou/NKtrader/NKtrader.java (hunk 1 above)
Comment (Darkvodou) [reply src/main/java/fr/darkvodou/NKtrader/NKtrader.java]: fixed next commit

File: src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java (hunk 19 above)
Comment (Darkvodou) [reply src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java]: no need world anymore

File: src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java (hunk 18 above)
Comment (Darkvodou) [reply src/main/java/fr/darkvodou/NKtrader/cmds/addTraderCmd.java]: moved verification in TraderManager.createTraderEntity(trader) or TraderManager.createTraderBlock(trader)

**Pull request merged**